data_api/snapshot_convertor/src/main/scala/uk/ac/wellcome/platform/snapshot_convertor/services/ConvertorService.scala

```diff
@@ -1,18 +1,85 @@
 package uk.ac.wellcome.platform.snapshot_convertor.services
 
 import javax.inject.Inject
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
 
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.Uri
+import akka.stream.ActorMaterializer
+import akka.stream.alpakka.s3.scaladsl.{MultipartUploadResult, S3Client}
+import akka.stream.scaladsl.Sink
+import akka.util.ByteString
 import com.twitter.inject.Logging
-import uk.ac.wellcome.platform.snapshot_convertor.models.ConversionJob
+import com.twitter.inject.annotations.Flag
+
+import uk.ac.wellcome.display.models.DisplayWork
+import uk.ac.wellcome.platform.snapshot_convertor.flow.{
+  ElasticsearchHitToDisplayWorkFlow,
+  StringToGzipFlow
+}
+import uk.ac.wellcome.platform.snapshot_convertor.models.{
+  CompletedConversionJob,
+  ConversionJob
+}
+import uk.ac.wellcome.platform.snapshot_convertor.source.S3Source
 import uk.ac.wellcome.utils.GlobalExecutionContext.context
+import uk.ac.wellcome.utils.JsonUtil._
 
-import scala.concurrent.Future
+class ConvertorService @Inject()(actorSystem: ActorSystem,
+                                 s3Client: S3Client,
+                                 @Flag("aws.s3.endpoint") s3Endpoint: String)
+    extends Logging {
 
-class ConvertorService @Inject()() extends Logging {
+  implicit val materializer = ActorMaterializer()(actorSystem)
 
-  def runConversion(conversionJob: ConversionJob): Future[Unit] = {
+  def runConversion(
+    conversionJob: ConversionJob): Future[CompletedConversionJob] = {
     info(s"ConvertorService running $conversionJob")
 
-    Future.failed(new RuntimeException("Not implemented!"))
+    val s3source = S3Source(
+      s3client = s3Client,
+      bucketName = conversionJob.bucketName,
+      key = conversionJob.objectKey
+    )
+
+    val displayWorks = s3source
+      .via(ElasticsearchHitToDisplayWorkFlow())
+
+    val jsonStrings = displayWorks
+      .map { displayWork: DisplayWork =>
+        toJson(displayWork)
+      }
+      .map {
+        case Success(jsonString) => jsonString
+        case Failure(encodeError) => {
+          warn("Failed to convert $displayWork to string!", encodeError)
+          throw encodeError
+        }
+      }
+
+    val gzipContent = jsonStrings
```

Comment:
This is a `Flow` (if I'm reading correctly). It could be useful to say that in the variable name. It would make things easier to follow at least for me. Something like `gzipContentFlow`. Same for `displayWorks` and `jsonStrings` above

Reply:
Yep. Sensible suggestion, will change.

Reply:
Change of plan – I’m going to add type annotations instead. Slightly clearer, and clarity enforced by the compiler! 😄

Reply:
👍